File: labs/07/src/Shape/Rectangle.php
<?php

namespace App\Shape;

use App\Canvas\CanvasInterface;
use App\Common\Point;

class Rectangle extends Shape
{
    final public function draw(CanvasInterface $canvas): void
    {
        $frame = $this->getFrame();

        $points = [
            new Point($frame->left, $frame->top),
            new Point($frame->left + $frame->width, $frame->top),
            new Point($frame->left + $frame->width, $frame->top + $frame->height),
            new Point($frame->left, $frame->top + $frame->height),
        ];

        $fillStyle = $this->getFillStyle();
        if ($fillStyle && $fillStyle->isEnabled() && $fillStyle->isEnabled() != null) {
            if ($fillStyle->getColor() !== null) {
                $canvas->setFillColor($fillStyle->getColor());
            }
            $canvas->fillRect($points);
        }

        $lineStyle = $this->getOutlineStyle();
        if ($lineStyle && $lineStyle->isEnabled() && $lineStyle->isEnabled() != null) {
            if ($lineStyle->getColor() !== null) {
                $canvas->setLineColor($lineStyle->getColor());
            }
            if ($lineStyle->getSize() !== null) {
                $canvas->setLineSize($lineStyle->getSize());
            }
            $canvas->drawRect($points);
        }
    }
}

File: labs/09/lab9v2/lab9v2/src/shape/Rectangle.java
package shape;

import canvas.CanvasInterface;

import java.awt.*;

public class Rectangle extends Shape {
    public Rectangle(Frame frame, Color strokeColor, Color fillColor) {
        super(frame, strokeColor, fillColor);
    }

    @Override
    public void draw(CanvasInterface canvas) {
        Frame frame = this.getFrame();
        Point[] points = {
            new Point(frame.getLeft(), frame.getTop()),
            new Point(frame.getLeft() + frame.getWidth(), frame.getTop()),
            new Point(frame.getLeft() + frame.getWidth(), frame.getTop() + frame.getHeight()),
            new Point(frame.getLeft(), frame.getTop() + frame.getHeight()),
        };
        canvas.fillPolygon(points, this.getFillStyle().getColor());
        canvas.setLineSize(3);
        canvas.setColor(this.getStrokeStyle().getColor());
        canvas.drawPolygon(points);
    }

    public boolean contains(Point point) {
        Frame frame = this.getFrame();
        int minX = frame.getLeft();
        int minY = frame.getTop();
        int maxX = frame.getLeft() + frame.getWidth();
        int maxY = frame.getTop() + frame.getHeight();

        return point.x >= minX && point.x <= maxX && point.y >= minY && point.y <= maxY;
    }

}
